components/tooltip/script.js

```diff
@@ -65,91 +78,114 @@ class Tooltip extends HTMLElement {
             case 'right':
-                position.left = elementSize.left + elementSize.width + TOOLTIP_MARGIN;
+                elementPosition.left = elementSize.left + elementSize.width + TOOLTIP_MARGIN;
                 break;
             default:
-                position.top = elementSize.top - TOOLTIP_MARGIN - tooltipSize.height;
+                elementPosition.top = elementSize.top - TOOLTIP_MARGIN - tooltipSize.height;
                 console.log(`The provided option for position ${orientation} is not valid - using top as a fallback. Possible options are top, bottom, left and right.`);
                 orientation = 'top';
                 break;
         }
 
-        this.position = orientation;
-
-        this.style.top = scrollOffsetY + position.top + 'px';
-        this.style.left = scrollOffsetX + position.left + 'px';
-
-        // check if the tooltip still overflows and re-position it again
-        const overflows  = this.overflows();
-        if(Object.keys(overflows).length && this.fixedSides.length !== 4) {
-            this.setPosition(this.getVisiblePosition(overflows));
-        }
+        return [orientation, elementPosition];
     }
 
-    overflows() {
-        var rect = this.getBoundingClientRect();
-        const overflows = {};
-
-        if (rect.top < 0) overflows.top = true;
-        if (rect.left < 0)  overflows.left = true;
-        if (rect.right > (window.innerWidth || document.documentElement.clientWidth)) overflows.right = true;
-        if (rect.bottom > (window.innerHeight || document.documentElement.clientHeight)) overflows.bottom = true;
+    /**
+     * Sets the new position of the tooltip. Waits 2 frames for the tooltip to be fully rendered
+     * and checks if the tooltip is visible on the new position. If it is not - calls setPosition
+     * recursively until a suitable position is found or we've ran out of possibilities.
+     * @param {number} scrollOffsetX - the offset of the scroll on the X axis.
+     * @param {number} scrollOffsetY - the offset of the scroll on the Y axis.
+     * @param {string} orientation - the current position of the tooltip - top, left, right or bottom.
+     * @returns {promise}
+    */
+    async setPosition(scrollOffsetX, scrollOffsetY, orientation = this.position) {
+        const elementSize = this.triggerElement.getBoundingClientRect();
 
-        return overflows;
-    }
+        let [updatedOrientation, elementPosition] = this.getPositionCoords(orientation, elementSize);
 
-    getVisiblePosition(overflows) {
-        const allSides = ['top', 'bottom', 'left', 'right'];
+        this.position = updatedOrientation;
+        this.style.top = scrollOffsetY + elementPosition.top + 'px';
+        this.style.left = scrollOffsetX + elementPosition.left + 'px';
 
-        overflows = overflows || this.overflows();
+        await this.waitForFrames(2);
+        const overflows = this.overflows;
         let overflowingSides = Object.keys(overflows);

```

Comment:
Why not do `Object.keys(this.overflows)` and remove the declaration from the previous row?

Reply:
Yes, removed.